File: Visual Studio 2010/CSWPAzureVideoStory/VideoStoryCreator/ViewModels/ComposePhotoViewModel.cs
/********************************* 模块头 *********************************\
* 模块名: ComposePhotoViewModel.cs
* 项目名: VideoStoryCreator
* 版权(c) Microsoft Corporation.
* 
* 原版页面的ViewModel类.
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Controls;
using VideoStoryCreator.Models;
using VideoStoryCreator.Transitions;

namespace VideoStoryCreator.ViewModels
{
    public class ComposePhotoViewModel : PhotoViewModel
    {
        public ITransition Transition { get; set; }

        private UserControl _transitionDesigner;
        public UserControl TransitionDesigner
        {
            get
            {
                return this._transitionDesigner;
            }
            set
            {
                if (this._transitionDesigner != value)
                {
                    this._transitionDesigner = value;
                    this.NotifyPropertyChange("TransitionDesigner");
                }
            }
        }

        public List<string> AvailableTransitions
        {
            get
            {
                return TransitionFactory.AvailableTransitions;
            }
        }

        /// <summary>
        ///  复制view model.
        /// </summary>
        /// <returns>view model备份.</returns>
        public ComposePhotoViewModel CopyTo()
        {
            ComposePhotoViewModel copy = new ComposePhotoViewModel()
            {
                Name = this.Name,
                PhotoDuration = this.PhotoDuration,
                TransitionDuration = this.TransitionDuration,
                MediaStream = this.MediaStream
            };
            if (this.Transition != null)
            {
                copy.Transition = this.Transition.Clone();
            }
            return copy;
        }

        /// <summary>
        /// 从复制的ViewModel类改变值.
        /// </summary>
        public void CopyFrom(ComposePhotoViewModel source)
        {
            this.Name = source.Name;
            this.PhotoDuration = source.PhotoDuration;
            this.TransitionDuration = source.TransitionDuration;
            if (source.Transition != null)
            {
                this.Transition = source.Transition.Clone();
            }
            this.MediaStream = source.MediaStream;
        }

        /// <summary>
        /// 将实体类改变为ViewModel类.
        /// </summary>
        public static ComposePhotoViewModel CreateFromModel(Photo model)
        {
            ComposePhotoViewModel viewModel = new ComposePhotoViewModel()
            {
                Name = model.Name,
                MediaStream = model.ThumbnailStream,
                PhotoDuration = (int)(model.PhotoDuration.TotalSeconds)                    
            };
            if (model.Transition != null)
            {
                viewModel.Transition = model.Transition;
                viewModel.TransitionDuration = (int)(model.Transition.TransitionDuration.TotalSeconds);
            }
            return viewModel;
        }

        /// <summary>
        /// 更行view model的通信.
        /// </summary>
        public void UpdateModel()
        {
            Photo photoToModify = App.MediaCollection.Where(p => p.Name == this.Name).FirstOrDefault();
            if (photoToModify != null && this.Transition != null)
            {
                photoToModify.Transition = this.Transition;
                photoToModify.Transition.TransitionDuration = TimeSpan.FromSeconds(this.TransitionDuration);
                photoToModify.PhotoDuration = TimeSpan.FromSeconds(this.PhotoDuration);
            }
        }

        /// <summary>
        /// 如果ViewModel被删除, 同样需要删除通信类.
        /// </summary>
        public void RemoveModel()
        {
            Photo model = App.MediaCollection.Where(p => p.Name == this.Name).FirstOrDefault();
            if (model != null)
            {
                model.ThumbnailStream.Close();
                App.MediaCollection.Remove(model);
            }
        }
    }
}

File: Visual Studio 2010/VBWPAzureVideoStory/VideoStoryCreator/ViewModels/ComposePhotoViewModel.vb
'********************************* 模块头 *********************************\
' 模块名: ComposePhotoViewModel.vb
' 项目名: VideoStoryCreator
' 版权(c) Microsoft Corporation.
' 
' 原版页面的ViewModel类.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'***************************************************************************/



Public Class ComposePhotoViewModel
    Inherits PhotoViewModel
    Public Property Transition() As ITransition
        Get
            Return m_Transition
        End Get
        Set(value As ITransition)
            m_Transition = value
        End Set
    End Property
    Private m_Transition As ITransition

    Private _transitionDesigner As UserControl
    Public Property TransitionDesigner() As UserControl
        Get
            Return Me._transitionDesigner
        End Get
        Set(value As UserControl)
            If Me._transitionDesigner IsNot value Then
                Me._transitionDesigner = value
                Me.NotifyPropertyChange("TransitionDesigner")
            End If
        End Set
    End Property

    Public ReadOnly Property AvailableTransitions() As List(Of String)
        Get
            Return TransitionFactory.AvailableTransitions
        End Get
    End Property

    ''' <summary>
    ''' 复制view model.
    ''' </summary>
    ''' <returns>view model备份.</returns>
    Public Function CopyTo() As ComposePhotoViewModel
        Dim copy As New ComposePhotoViewModel() With { _
          .Name = Me.Name, _
          .PhotoDuration = Me.PhotoDuration, _
          .TransitionDuration = Me.TransitionDuration, _
          .MediaStream = Me.MediaStream _
        }
        If Me.Transition IsNot Nothing Then
            copy.Transition = Me.Transition.Clone()
        End If
        Return copy
    End Function

    ''' <summary>
    ''' 从复制的ViewModel类改变值.
    ''' </summary>
    Public Sub CopyFrom(source As ComposePhotoViewModel)
        Me.Name = source.Name
        Me.PhotoDuration = source.PhotoDuration
        Me.TransitionDuration = source.TransitionDuration
        If source.Transition IsNot Nothing Then
            Me.Transition = source.Transition.Clone()
        End If
        Me.MediaStream = source.MediaStream
    End Sub

    ''' <summary>
    ''' 将实体类改变为ViewModel类.
    ''' </summary>
    Public Shared Function CreateFromModel(model As Photo) As ComposePhotoViewModel
        Dim viewModel As New ComposePhotoViewModel() With { _
          .Name = model.Name, _
          .MediaStream = model.ThumbnailStream, _
          .PhotoDuration = CInt(model.PhotoDuration.TotalSeconds) _
        }
        If model.Transition IsNot Nothing Then
            viewModel.Transition = model.Transition
            viewModel.TransitionDuration = CInt(model.Transition.TransitionDuration.TotalSeconds)
        End If
        Return viewModel
    End Function

    ''' <summary>
    ''' 更行view model的通信.
    ''' </summary>
    Public Sub UpdateModel()
        Dim photoToModify As Photo = App.MediaCollection.Where(Function(p) p.Name = Me.Name).FirstOrDefault()
        If photoToModify IsNot Nothing AndAlso Me.Transition IsNot Nothing Then
            photoToModify.Transition = Me.Transition
            photoToModify.Transition.TransitionDuration = TimeSpan.FromSeconds(Me.TransitionDuration)
            photoToModify.PhotoDuration = TimeSpan.FromSeconds(Me.PhotoDuration)
        End If
    End Sub

    ''' <summary>
    '''  如果ViewModel被删除, 同样需要删除通信类.
    ''' </summary>
    Public Sub RemoveModel()
        Dim model As Photo = App.MediaCollection.Where(Function(p) p.Name = Me.Name).FirstOrDefault()
        If model IsNot Nothing Then
            model.ThumbnailStream.Close()
            App.MediaCollection.Remove(model)
        End If
    End Sub
End Class
